### PR TITLE
Use UCC for send, and recv for ProcessGroupNCCL on CPU

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR is based on [#69552](https://github.com/pytorch/pytorch/pull/69564), please review that PR first.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -1591,7 +1603,7 @@ def test_isend_torch_profiler(self):
             self._test_isend(profiler_ctx=torch_profiler_ctx)
 
         # IRECV
-        @sandcastle_skip_if(BACKEND == "nccl", "Nccl does not support irecv")
+        @require_ucc_for_nccl()
```
Comment (dzhulgakov) [inline torch/testing/_internal/distributed/distributed_test.py:1606]: keep it `sandcastle_skip_if` and just make `require_ucc_for_nccl` be a function returning bool

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -376,6 +377,23 @@ def require_backend(backends):
...
 
+def require_ucc_for_nccl():
+    def do_skip():
+        def is_ucc_available():
+            return dist.distributed_c10d._get_default_group().is_ucc_available
+
+        if BACKEND == dist.Backend.NCCL and not is_ucc_available():
+            raise unittest.SkipTest("Test requires UCC to be available for the NCCL backend")
+
+    def wrapper(func):
```
Comment (dzhulgakov) [inline torch/testing/_internal/distributed/distributed_test.py:388]: no need to do it if you use `skip_if`

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -589,6 +590,35 @@ ProcessGroupNCCL::ProcessGroupNCCL(
...
+    uccLib_ = loadTorchUCC();
+    if (uccLib_ != nullptr) {
+      LOG(INFO) << "[Rank " << rank_  << "] torch_ucc.so loaded";
+    }
+  });
+
+  if (uccLib_ != nullptr) {
+    LOG(INFO) << "[Rank " << rank_  << "] torch_ucc.so loaded";
+    typedef void *fn(void *);
+    auto createProcessGroupUCCForNCCL = reinterpret_cast<fn*>(uccLib_->sym("_Z28createProcessGroupUCCForNCCLPv"));
```
Comment (dzhulgakov) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:606]: use export "C", you won't need mangling

**Review by dzhulgakov — COMMENTED** [review]
I wonder whether compiler time flag adds any value on top of runtime check.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (zasdfgbnm) [reply torch/testing/_internal/distributed/distributed_test.py:393]: I can not use `skip_if`. In order to use `skip_if`, the condition needs to be evaluated at the test-discovery time. And at that time, the default group is not initialized yet, so I don't have `is_ucc_available` available. I have to defer the evaluation of `is_ucc_available` to the time when test runs.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (zasdfgbnm) [reply torch/testing/_internal/distributed/distributed_test.py:1791]: See https://github.com/pytorch/pytorch/pull/69587/files#r850709419

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (zasdfgbnm) [reply torch/testing/_internal/distributed/distributed_test.py:393]: Maybe alternatively, I can change the skip condition to "`TORCH_UCC_LIBRARY_PATH` is set"

**Comment by zasdfgbnm** [comment]
This is mostly ready, please land https://github.com/facebookresearch/torch_ucc/pull/70 first before landing this PR.

**Pull request closed without merging**